### PR TITLE
[RHELC-1037] Automatic Manpage Generation with Pre-Commit

### PR DESCRIPTION
**Pull request by Andrew-ang9**
This PR introduces a new approach to manpage generation by leveraging the pre-commit framework. 

Jira Issues: [RHELC-1037](https://issues.redhat.com/browse/RHELC-1037)

Checklist

- [ ] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [ ] `[RHELC-]` is part of the PR title 
- [ ] GitHub label has been added to help with Release notes 
- [ ] PR title explains the change from the user's point of view
- [ ] Code and tests are documented properly
- [ ] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

**Comment by r0x0d** [comment]
Tried it locally, but no manpage file was generated.

**Comment by Andrew-ang9** [comment]
> Tried it locally, but no manpage file was generated.

I will look into this issue! Thank you for the help!

File: .pre-commit-config.yaml
Comment (SpyTec) [inline .pre-commit-config.yaml:69]: We might just wanna wrap the manpage generation in a container if possible so it works on all machines

Comment (Andrew-ang9) [reply .pre-commit-config.yaml:69]: Ok, I can do that! I'll work on that and get it pushed back up to this PR!

File: scripts/manpage_generation.sh
Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:1]: ahh ok I have already put it in a python file and I was looking into getting the dependencies but had no luck. I will but i was not wraping it in a containerfile so I will do that now!

File: man/convert2rhel.8
Comment (Andrew-ang9) [reply man/convert2rhel.8:1]: gotcha thank you for the reference!

File: scripts/manpage_generation.sh
Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:1]: @SpyTec Just to be sure I should make a Containerfile for the manpages right?

Comment (SpyTec) [reply scripts/manpage_generation.sh:1]: Right, run the manpage generation inside a container and make it work locally. Then we should be able to execute it in pre-commit

**Comment by Andrew-ang9** [comment]
I have put this PR ready for review now as I have finished making the container file for it. I am now trying to get podman configured so that I can run the container, but I have got it mostly done and want to just check to see if I set up the container file up right.

File: scripts/manpage_generation.sh
Comment (Venefilyn) [inline scripts/manpage_generation.sh:15]: Good solution

```suggestion
# Detect changes in __init__.py
if git diff --quiet HEAD~1 -- convert2rhel/__init__.py; then
    # No changes detected, exit with 0
    echo "No new version detected"
    exit 0
else
    echo "New version detected"
    # Changes detected, generate manpages
```

File: scripts/manpage_generation.sh
Comment (Venefilyn) [inline scripts/manpage_generation.sh:40]: This is fine for now but we might need to modify this to support multiple versions soon, so we can support RHEL 9. Out of scope for this PR though

File: scripts/manpage_generation.sh
Comment (Venefilyn) [inline scripts/manpage_generation.sh:17]:
```suggestion
    #get dependency
```

File: scripts/manpage_generation.sh
Comment (Venefilyn) [inline scripts/manpage_generation.sh:21]: This can be moved out to a `requirements.txt` file, can get inspiration from other requirement files in the repo. To get it from the container directly you can do `pip freeze -f requirements.txt`. Then add this as a step in the container, this way we can cache it

Comment (Venefilyn) [reply scripts/manpage_generation.sh:21]: Action is also failing here due to syntax error

https://github.com/oamg/convert2rhel/actions/runs/9072412663/job/24927859923

**Review by Venefilyn — CHANGES_REQUESTED** [review]
Seems like a good solution! Some comments

File: scripts/manpage_generation.sh
Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:15]: That was a template for when I was working on this. I can get that changed!

File: scripts/manpage_generation.sh
Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:40]: Ok, should I make an Jira issue for that after this one is merged?

File: scripts/manpage_generation.sh
Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:21]: Gotcha I can get that moved out to an requirements.txt file! I will also get the action fixed too!

Thank you for the help with theses changes!

File: scripts/manpage_generation.sh
Comment (Venefilyn) [reply scripts/manpage_generation.sh]: I think we can leave it for now

File: .pre-commit-config.yaml
Comment (Venefilyn) [inline .pre-commit-config.yaml:63]: Did you test this locally? It doesn't work

```
convert2rhel on  pr/Andrew-ang9/1106:updated_fix_manpage [$] via 🐍 v3.12.3 
❯ podman run --rm Containerfiles/manpage.Containerfile          
Error: repository name must be lowercase
```
You could probably do this instead. Where `podman build`'s `-q` is quiet and will only output the image hash, which will then execute to run as `podman run --rm <image_hash>`
```suggestion
        entry: podman run --rm $(podman build -q -f Containerfiles/manpage.Containerfile)
```

File: Containerfiles/manpage.Containerfile
Comment (Venefilyn) [inline Containerfiles/manpage.Containerfile:23]: Tried running the container as is first, and was met with this
```
❯ podman run --rm $(podman build -q -f Containerfiles/manpage.Containerfile)
Error: Unable to find a match: requirements/manpages_requirements.txt
Error: building at STEP "RUN yum install -y requirements/manpages_requirements.txt": while running runtime: exit status 1
Error: requires at least 1 arg(s), only received 0
```

**Review by Venefilyn — CHANGES_REQUESTED** [review]

File: Containerfiles/manpage.Containerfile
Comment (Andrew-ang9) [reply Containerfiles/manpage.Containerfile:23]: I haven gotten to this error, as of yet as I have been trying to get my container working. 

I will continue to look at this and see what I can do to get it working

File: .pre-commit-config.yaml
Comment (Andrew-ang9) [reply .pre-commit-config.yaml:63]: Yeah I have done some testing locally and ran into some issues. I asked for the review anyways to see if I could get a little help with this. Thank you!

File: scripts/manpage_generation.py
Comment (r0x0d) [inline scripts/manpage_generation.py:31]:
```suggestion
    subprocess.run(["dnf", "install", "-y", "python3-rpm", "python3-dnf"], check=True)
```

Comment (Andrew-ang9) [reply scripts/manpage_generation.py:31]: Thank you for the extra look on this, I will get this taken care of and push it up!

File: scripts/manpage_generation.py
Comment (r0x0d) [inline scripts/manpage_generation.py:32]:
```suggestion
```

File: scripts/manpage_generation.py
Comment (r0x0d) [inline scripts/manpage_generation.py:30]: You could install this as part of the container as well

File: scripts/manpage_generation.sh
Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:1]: @Venefilyn So we don't need this as its own GitHub action and it can just be ran within the pre-commit action and locally with pre-commit right?

Comment (Venefilyn) [reply scripts/manpage_generation.sh:1]: Yeah that's what I'm thinking!

Comment (Andrew-ang9) [reply scripts/manpage_generation.sh:1]: Ok, cool that's what I was thinking as well but I wanted to confirm. I am cleaning up this PR some and noticed that. I am also working on getting pre-commit to pass and making sure that the manpages are actually generated!

File: scripts/manpage_generation.py
Comment (Venefilyn) [inline scripts/manpage_generation.py:27]: I'd say comment this out and have the script always generate manpages

File: .pre-commit-config.yaml
Comment (Venefilyn) [inline .pre-commit-config.yaml:66]: Actually, this might be difficult to do. I'm not sure but I couldn't get this to work locally. Maybe we just stick to the GitHub workflow. But for that to work we need to ensure that `podman run --rm $(podman build -q -f Containerfiles/manpage.Containerfile)` works locally

**Pull request closed without merging**
